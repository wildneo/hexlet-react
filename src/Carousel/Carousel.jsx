/* eslint-disable react/prop-types */
import React from 'react';
import cn from 'classnames';

export default class Carousele extends React.Component {
  constructor(props) {
    super(props);
    const maxIndex = props.images.length;
    this.state = {
      active: 0,
      maxIndex,
    };

    this.next = () => this.increment();
    this.prev = () => this.decrement();
  }

  increment() {
    const { active, maxIndex } = this.state;
    this.setState({ active: (active + 1) % maxIndex });
  }

  decrement() {
    const { active, maxIndex } = this.state;
    this.setState({ active: active === 0 ? maxIndex - 1 : (active - 1) });
  }

  render() {
    const { images } = this.props;

    const process = (image, key) => {
      const { active } = this.state;
      const style = cn('carousel-item', { active: active === key });
      return (
        <div key={key} className={style}>
          <img alt="" className="d-block w-100" src={image} />
        </div>
      );
    };

    return (
      <div id="carousel" className="carousel slide" data-ride="carousel">
        <div className="carousel-inner">{images.map(process)}</div>
        <a onClick={this.prev} className="carousel-control-prev" href="#carousel" role="button" data-slide="prev">
          <span className="carousel-control-prev-icon" />
          <span className="sr-only">Previous</span>
        </a>
        <a onClick={this.next} className="carousel-control-next" href="#carousel" role="button" data-slide="next">
          <span className="carousel-control-next-icon" />
          <span className="sr-only">Next</span>
        </a>
      </div>
    );
  }
}
