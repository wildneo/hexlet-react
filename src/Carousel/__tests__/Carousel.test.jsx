import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Carousel from '../Carousel';

Enzyme.configure({ adapter: new Adapter() });

test('Carousel', () => {
  const images = ['/images/first.jpeg', '/images/second.jpeg', '/images/third.jpeg'];
  const wrapper = mount(<Carousel images={images} />);
  expect(wrapper.render()).toMatchSnapshot();

  const next = wrapper.find('.carousel-control-next');
  const prev = wrapper.find('.carousel-control-prev');

  next.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  next.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  next.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  prev.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  prev.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  prev.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  prev.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
});

test('Carousel 2', () => {
  const images = ['/images/1.jpeg', '/images/2.jpeg', '/images/3.jpeg'];
  const wrapper = mount(<Carousel images={images} />);
  expect(wrapper.render()).toMatchSnapshot();

  const next = wrapper.find('.carousel-control-next');

  next.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  next.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  next.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
});
