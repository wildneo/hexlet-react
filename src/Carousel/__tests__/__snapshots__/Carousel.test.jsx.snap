// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel 1`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/first.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/second.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/third.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 2`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/first.jpeg"
      />
    </div>
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/second.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/third.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 2 1`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/1.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/2.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/3.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 2 2`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/1.jpeg"
      />
    </div>
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/2.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/3.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 2 3`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/1.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/2.jpeg"
      />
    </div>
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/3.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 2 4`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/1.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/2.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/3.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 3`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/first.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/second.jpeg"
      />
    </div>
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/third.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 4`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/first.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/second.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/third.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 5`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/first.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/second.jpeg"
      />
    </div>
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/third.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 6`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/first.jpeg"
      />
    </div>
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/second.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/third.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 7`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/first.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/second.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/third.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;

exports[`Carousel 8`] = `
<div
  class="carousel slide"
  data-ride="carousel"
  id="carousel"
>
  <div
    class="carousel-inner"
  >
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/first.jpeg"
      />
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/second.jpeg"
      />
    </div>
    <div
      class="carousel-item active"
    >
      <img
        alt=""
        class="d-block w-100"
        src="/images/third.jpeg"
      />
    </div>
  </div>
  <a
    class="carousel-control-prev"
    data-slide="prev"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    data-slide="next"
    href="#carousel"
    role="button"
  >
    <span
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;
